File: go/spore-eval/metricmap/metricmap.go
// Package metricmap defines the EvalMetric type + extraction of metric samples
// from observability (Rule 16, Rule 17, Resolution 1).
//
// Simple aggregates (turns, cost, wall-time, success) come from SessionMetrics
// via GetSessionMetrics. Filtered metrics (CacheHitRate{block},
// SensorFireRate{sensor_id}, MiddlewareInterventionRate{hook}) are computed
// from GetTrace spans (Resolution 1).
//
// Per Resolution 1 (Go note): the three span-sourced metrics are read via a
// clean type switch on the concrete span payload types (TurnSpan,
// SensorSpan, MiddlewareSpan) — NOT via the Debug-string-parsing hack the Rust
// reference uses as a workaround. The ObservabilityProvider public surface is
// not modified.
package metricmap

import (
	"fmt"

	sporecore "github.com/squirrelsoft-dev/spore-core/go/spore-core"
	"github.com/squirrelsoft-dev/spore-core/go/spore-core/middleware"
	"github.com/squirrelsoft-dev/spore-core/go/spore-core/observability"
)

// EvalMetricKind discriminates EvalMetric variants.
type EvalMetricKind string

const (
	// MetricTaskSuccessRate — fraction of runs whose verifier passed.
	MetricTaskSuccessRate EvalMetricKind = "task_success_rate"
	// MetricMeanTurnsToCompletion — mean total turns.
	MetricMeanTurnsToCompletion EvalMetricKind = "mean_turns_to_completion"
	// MetricMeanCostUsd — mean total cost in USD.
	MetricMeanCostUsd EvalMetricKind = "mean_cost_usd"
	// MetricMeanWallTime — mean wall time (ms).
	MetricMeanWallTime EvalMetricKind = "mean_wall_time"
	// MetricCacheHitRate — cache-read hit rate for a named cache block.
	MetricCacheHitRate EvalMetricKind = "cache_hit_rate"
	// MetricSensorFireRate — fire rate of a named sensor.
	MetricSensorFireRate EvalMetricKind = "sensor_fire_rate"
	// MetricMiddlewareInterventionRate — intervention rate of a named hook.
	MetricMiddlewareInterventionRate EvalMetricKind = "middleware_intervention_rate"
	// MetricVerificationScore — mean verification score across runs.
	MetricVerificationScore EvalMetricKind = "verification_score"
)

// EvalMetric is a metric the EvalHarness aggregates and compares (Rule 17).
// Block / SensorID / Hook carry the parameter for the filtered variants.
type EvalMetric struct {
	Kind EvalMetricKind `json:"kind"`
	// cache_hit_rate
	Block string `json:"-"`
	// sensor_fire_rate
	SensorID string `json:"-"`
	// middleware_intervention_rate
	Hook string `json:"-"`
}

// Constructors for the simple variants.

// TaskSuccessRate is the success-rate metric.
func TaskSuccessRate() EvalMetric { return EvalMetric{Kind: MetricTaskSuccessRate} }

// VerificationScore is the mean verification-score metric.
func VerificationScore() EvalMetric { return EvalMetric{Kind: MetricVerificationScore} }

// MeanTurns is the mean-turns metric.
func MeanTurns() EvalMetric { return EvalMetric{Kind: MetricMeanTurnsToCompletion} }

// MeanCost is the mean-cost metric.
func MeanCost() EvalMetric { return EvalMetric{Kind: MetricMeanCostUsd} }

// MeanWallTime is the mean-wall-time metric.
func MeanWallTime() EvalMetric { return EvalMetric{Kind: MetricMeanWallTime} }

// CacheHitRate is the cache-hit-rate metric for a named block.
func CacheHitRate(block string) EvalMetric {
	return EvalMetric{Kind: MetricCacheHitRate, Block: block}
}

// SensorFireRate is the sensor-fire-rate metric for a named sensor.
func SensorFireRate(sensorID string) EvalMetric {
	return EvalMetric{Kind: MetricSensorFireRate, SensorID: sensorID}
}

// MiddlewareInterventionRate is the intervention-rate metric for a named hook.
func MiddlewareInterventionRate(hook string) EvalMetric {
	return EvalMetric{Kind: MetricMiddlewareInterventionRate, Hook: hook}
}

// MarshalJSON serialises as a flat tagged object keyed by "kind".
func (m EvalMetric) MarshalJSON() ([]byte, error) {
	switch m.Kind {
	case MetricCacheHitRate:
		return jsonMarshal(struct {
			Kind  EvalMetricKind `json:"kind"`
			Block string         `json:"block"`
		}{m.Kind, m.Block})
	case MetricSensorFireRate:
		return jsonMarshal(struct {
			Kind     EvalMetricKind `json:"kind"`
			SensorID string         `json:"sensor_id"`
		}{m.Kind, m.SensorID})
	case MetricMiddlewareInterventionRate:
		return jsonMarshal(struct {
			Kind EvalMetricKind `json:"kind"`
			Hook string         `json:"hook"`
		}{m.Kind, m.Hook})
	default:
		return jsonMarshal(struct {
			Kind EvalMetricKind `json:"kind"`
		}{m.Kind})
	}
}

// UnmarshalJSON decodes the flat tagged form.
func (m *EvalMetric) UnmarshalJSON(data []byte) error {
	var probe struct {
		Kind     EvalMetricKind `json:"kind"`
		Block    string         `json:"block"`
		SensorID string         `json:"sensor_id"`
		Hook     string         `json:"hook"`
	}
	if err := jsonUnmarshal(data, &probe); err != nil {
		return err
	}
	m.Kind = probe.Kind
	m.Block = probe.Block
	m.SensorID = probe.SensorID
	m.Hook = probe.Hook
	if m.Kind == "" {
		return fmt.Errorf("EvalMetric: missing kind")
	}
	return nil
}

// Direction returns the optimization direction: higher-is-better metrics
// maximize; cost, turns, wall-time, and intervention/sensor rates minimize.
func (m EvalMetric) Direction() sporecore.OptimizationDirection {
	switch m.Kind {
	case MetricTaskSuccessRate, MetricCacheHitRate, MetricVerificationScore:
		return sporecore.OptimizationMaximize
	default:
		return sporecore.OptimizationMinimize
	}
}

// Name returns a stable display/serialization name.
func (m EvalMetric) Name() string {
	switch m.Kind {
	case MetricTaskSuccessRate:
		return "task_success_rate"
	case MetricMeanTurnsToCompletion:
		return "mean_turns_to_completion"
	case MetricMeanCostUsd:
		return "mean_cost_usd"
	case MetricMeanWallTime:
		return "mean_wall_time"
	case MetricCacheHitRate:
		return fmt.Sprintf("cache_hit_rate[%s]", m.Block)
	case MetricSensorFireRate:
		return fmt.Sprintf("sensor_fire_rate[%s]", m.SensorID)
	case MetricMiddlewareInterventionRate:
		return fmt.Sprintf("middleware_intervention_rate[%s]", m.Hook)
	case MetricVerificationScore:
		return "verification_score"
	default:
		return string(m.Kind)
	}
}

// FromTrace reports whether this metric is sourced from filtered trace spans
// (Resolution 1) rather than from the SessionMetrics aggregate.
func (m EvalMetric) FromTrace() bool {
	switch m.Kind {
	case MetricCacheHitRate, MetricSensorFireRate, MetricMiddlewareInterventionRate:
		return true
	default:
		return false
	}
}

// RunSampleInputs are the per-run inputs not in observability: the verifier
// outcome.
type RunSampleInputs struct {
	VerifierPassed bool
	VerifierScore  float64
}

// SampleFor computes the sample value of metric for a single run, given that
// run's SessionMetrics (simple aggregates), the run's trace spans (filtered
// metrics, Resolution 1), and the verifier outcome.
func SampleFor(metric EvalMetric, session *observability.SessionMetrics, spans []observability.Span, inputs RunSampleInputs) float64 {
	switch metric.Kind {
	case MetricTaskSuccessRate:
		if inputs.VerifierPassed {
			return 1.0
		}
		return 0.0
	case MetricVerificationScore:
		return inputs.VerifierScore
	case MetricMeanTurnsToCompletion:
		return float64(session.TotalTurns)
	case MetricMeanCostUsd:
		return session.TotalCostUSD
	case MetricMeanWallTime:
		return float64(session.TotalDurationMs)
	case MetricCacheHitRate:
		return cacheHitRate(spans)
	case MetricSensorFireRate:
		return sensorFireRate(spans, metric.SensorID)
	case MetricMiddlewareInterventionRate:
		return middlewareInterventionRate(spans, metric.Hook)
	default:
		return 0.0
	}
}

// cacheHitRate is cache_read_tokens / (input + cache_read) summed over turn
// spans (Resolution 1, TurnSpan.CacheReadTokens). 0.0 with no tokens.
func cacheHitRate(spans []observability.Span) float64 {
	var cacheRead, input uint64
	for _, s := range spans {
		if ts, ok := s.(observability.TurnSpan); ok {
			if ts.CacheReadTokens != nil {
				cacheRead += uint64(*ts.CacheReadTokens)
			}
			input += uint64(ts.InputTokens)
		}
	}
	denom := input + cacheRead
	if denom == 0 {
		return 0.0
	}
	return float64(cacheRead) / float64(denom)
}

// sensorFireRate is fired evaluations / total evaluations of the named sensor
// (Resolution 1, SensorSpan.SensorID). 0.0 if never evaluated.
func sensorFireRate(spans []observability.Span, sensorID string) float64 {
	var fired, total uint32
	for _, s := range spans {
		if ss, ok := s.(observability.SensorSpan); ok {
			if string(ss.SensorID) == sensorID {
				total++
				if ss.Fired {
					fired++
				}
			}
		}
	}
	if total == 0 {
		return 0.0
	}
	return float64(fired) / float64(total)
}

// middlewareInterventionRate is non-Continue decisions / total firings at the
// named hook (Resolution 1, MiddlewareSpan.Hook).
func middlewareInterventionRate(spans []observability.Span, hook string) float64 {
	var interventions, total uint32
	for _, s := range spans {
		if ms, ok := s.(observability.MiddlewareSpan); ok {
			if string(ms.Hook) == hook {
				total++
				if ms.Decision.Kind != middleware.DecisionContinue {
					interventions++
				}
			}
		}
	}
	if total == 0 {
		return 0.0
	}
	return float64(interventions) / float64(total)
}

File: python/packages/spore_eval/src/spore_eval/eval/metric_map.py
""":data:`EvalMetric` tagged union + extraction of metric samples from
observability (Rule 16, Rule 17, Resolution 1).

Mirrors ``rust/crates/spore-eval/src/metric_map.rs`` in shape and rules, but
NOT in implementation: the simple aggregates (turns, cost, wall-time, success)
come from :class:`SessionMetrics` via ``get_session_metrics``, and the filtered
metrics (``CacheHitRate{block}``, ``SensorFireRate{sensor_id}``,
``MiddlewareInterventionRate{hook}``) are computed from ``get_trace`` spans by
reading the **typed span attributes directly** (Resolution 1 NOTE) — Python
does not replicate the Rust Debug-string-parsing workaround.

Issue #12's public surface is NOT modified.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Annotated, Literal

from pydantic import BaseModel, ConfigDict, Field
from spore_core.harness import OptimizationDirection
from spore_core.middleware import MiddlewareContinue
from spore_core.observability import (
    MiddlewareSpan,
    SensorSpan,
    SessionMetrics,
    Span,
    TurnSpan,
)


class _Model(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)


# ============================================================================
# EvalMetric — tagged union on ``kind`` (Rule 17)
# ============================================================================


class EvalMetricTaskSuccessRate(_Model):
    kind: Literal["task_success_rate"] = "task_success_rate"


class EvalMetricMeanTurnsToCompletion(_Model):
    kind: Literal["mean_turns_to_completion"] = "mean_turns_to_completion"


class EvalMetricMeanCostUsd(_Model):
    kind: Literal["mean_cost_usd"] = "mean_cost_usd"


class EvalMetricMeanWallTime(_Model):
    kind: Literal["mean_wall_time"] = "mean_wall_time"


class EvalMetricCacheHitRate(_Model):
    kind: Literal["cache_hit_rate"] = "cache_hit_rate"
    block: str


class EvalMetricSensorFireRate(_Model):
    kind: Literal["sensor_fire_rate"] = "sensor_fire_rate"
    sensor_id: str


class EvalMetricMiddlewareInterventionRate(_Model):
    kind: Literal["middleware_intervention_rate"] = "middleware_intervention_rate"
    hook: str


class EvalMetricVerificationScore(_Model):
    kind: Literal["verification_score"] = "verification_score"


EvalMetric = Annotated[
    EvalMetricTaskSuccessRate
    | EvalMetricMeanTurnsToCompletion
    | EvalMetricMeanCostUsd
    | EvalMetricMeanWallTime
    | EvalMetricCacheHitRate
    | EvalMetricSensorFireRate
    | EvalMetricMiddlewareInterventionRate
    | EvalMetricVerificationScore,
    Field(discriminator="kind"),
]

_MAXIMIZE: set[str] = {"task_success_rate", "cache_hit_rate", "verification_score"}


def metric_direction(metric: EvalMetric) -> OptimizationDirection:
    """The optimization direction: higher-is-better metrics maximize; cost,
    turns, wall-time, sensor-fire and intervention rates minimize."""
    return "maximize" if metric.kind in _MAXIMIZE else "minimize"


def metric_name(metric: EvalMetric) -> str:
    """A stable display/serialization name."""
    if isinstance(metric, EvalMetricCacheHitRate):
        return f"cache_hit_rate[{metric.block}]"
    if isinstance(metric, EvalMetricSensorFireRate):
        return f"sensor_fire_rate[{metric.sensor_id}]"
    if isinstance(metric, EvalMetricMiddlewareInterventionRate):
        return f"middleware_intervention_rate[{metric.hook}]"
    return metric.kind


def metric_from_trace(metric: EvalMetric) -> bool:
    """Whether this metric is sourced from filtered trace spans (Resolution 1)
    rather than from the simple :class:`SessionMetrics` aggregate."""
    return isinstance(
        metric,
        (
            EvalMetricCacheHitRate,
            EvalMetricSensorFireRate,
            EvalMetricMiddlewareInterventionRate,
        ),
    )


# ============================================================================
# RunSampleInputs + sample_for
# ============================================================================


@dataclass
class RunSampleInputs:
    """Per-run inputs needed to derive a metric sample that are not in
    observability: the verifier outcome."""

    verifier_passed: bool
    verifier_score: float


def sample_for(
    metric: EvalMetric,
    session: SessionMetrics,
    spans: list[Span],
    inputs: RunSampleInputs,
) -> float:
    """Compute the sample value of ``metric`` for a single run, given that
    run's :class:`SessionMetrics` (simple aggregates), the run's trace
    ``spans`` (filtered metrics, Resolution 1), and the verifier outcome."""
    if isinstance(metric, EvalMetricTaskSuccessRate):
        return 1.0 if inputs.verifier_passed else 0.0
    if isinstance(metric, EvalMetricVerificationScore):
        return inputs.verifier_score
    if isinstance(metric, EvalMetricMeanTurnsToCompletion):
        return float(session.total_turns)
    if isinstance(metric, EvalMetricMeanCostUsd):
        return session.total_cost_usd
    if isinstance(metric, EvalMetricMeanWallTime):
        return float(session.total_duration_ms)
    if isinstance(metric, EvalMetricCacheHitRate):
        return _cache_hit_rate(spans)
    if isinstance(metric, EvalMetricSensorFireRate):
        return _sensor_fire_rate(spans, metric.sensor_id)
    if isinstance(metric, EvalMetricMiddlewareInterventionRate):
        return _middleware_intervention_rate(spans, metric.hook)
    raise AssertionError(f"unhandled metric {metric!r}")  # pragma: no cover


# ---- Resolution 1: read typed span attributes directly (no Debug parsing) ---


def _cache_hit_rate(spans: list[Span]) -> float:
    """Cache-read hit rate from turn spans:
    ``cache_read / (input + cache_read)``. 0.0 with no tokens."""
    cache_read = 0
    input_tokens = 0
    for s in spans:
        if isinstance(s, TurnSpan):
            cache_read += s.cache_read_tokens or 0
            input_tokens += s.input_tokens
    denom = input_tokens + cache_read
    return cache_read / denom if denom else 0.0


def _sensor_fire_rate(spans: list[Span], sensor_id: str) -> float:
    """Fire rate of a named sensor: fired / total evaluations of that sensor.
    0.0 if never evaluated."""
    fired = 0
    total = 0
    for s in spans:
        if isinstance(s, SensorSpan) and str(s.sensor_id) == sensor_id:
            total += 1
            if s.fired:
                fired += 1
    return fired / total if total else 0.0


def _middleware_intervention_rate(spans: list[Span], hook: str) -> float:
    """Intervention rate of a named middleware hook: non-Continue decisions /
    total firings at that hook."""
    interventions = 0
    total = 0
    for s in spans:
        if isinstance(s, MiddlewareSpan) and _hook_name(s.hook).lower() == hook.lower():
            total += 1
            if not isinstance(s.decision, MiddlewareContinue):
                interventions += 1
    return interventions / total if total else 0.0


def _hook_name(hook: object) -> str:
    """The hook's stable string name (``HookPoint`` is a ``str`` enum)."""
    value = getattr(hook, "value", None)
    return value if isinstance(value, str) else str(hook)
